docs/spec_draft.md

```diff
@@ -1810,6 +1811,83 @@ where `i[d] = j[permutation[d]]`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.triangular_solve
+
+### Semantics
+
+Solves systems of linear equations of a batch of matrices with lower or upper
+triangular coefficients by forward- or back-substitution.
+
+More formally, for all `i`, `result[i0, ..., iR-3, :, :]` is the solution to
+`op(a[i0, ..., iR-3, :, :]) * x = b[i0, ..., iR-3, :, :]` (`left_side` is true)
+`x * op(a[i0, ..., iR-3, :, :]) = b[i0, ..., iR-3, :, :]` (`left_side` is false)
+solving for the variable `x` given `a` and `b`, where `op(a)` is either
+`op(a) = a`, or `op(a) = Transpose(a)`, or `op(a) = Conj(Transpose(a))`. The
+output values in the opposite triangle, i.e. the strict upper triangle or strict
+lower triangle correspondingly, are implementation-defined.
+
+If `unit_diagonal` is true, the diamonal elements of `a` are assumed to be 1 and
+not accessed.
+
+`transpose_a` can be one of the following:
+  * `TRANSPOSE_INVALID`: Illegal matrix; transpose is invalid.
+  * `NO_TRANSPOSE`: Perform operation as-is.
+  * `TRANSPOSE`: Perform operation on transpose of `a`.
+  * `ADJOINT`: Perform operation on conjugate transpose of `a`.
+
+### Inputs
+
+| Name            | Type                                                                    |
+|-----------------|-------------------------------------------------------------------------|
+| `a`             | tensor of floating-point or complex type                                |
+| `b`             | tensor of floating-point or complex type                                |
+| `left_side`     | constant of type `i1`                                                   |
+| `lower`         | constant of type `i1`                                                   |
+| `unit_diagonal` | constant of type `i1`                                                   |
+| `transpose_a`   | enum of `TRANSPOSE_INVALID`, `NO_TRANSPOSE`, `TRANSPOSE`, and `ADJOINT` |
+
+### Outputs
+
+| Name     | Type                                     |
+|----------|------------------------------------------|
+| `result` | tensor of floating-point or complex type |
+
+### Constraints
+
+  * (C1) `a` and `b` have the same type and rank $\ge$ 2.
+  * (C2) dim(`a`, -2) = dim(`a`, -1).
+  * (C3) `a` and `b` have same shape except `R-1` and `R-2` dimensions.
+  * (C4) Suppose `a` has shape `[..., M, M]`, then `b` has shape `[..., M, K]`
```

Comment:
* dim(`a`, i) $=$ dim(`b`, i) if i $\in$ [0, R-3] and dim(`a`, i) $=$ dim(`b`, R - `left_side` ? 2 : 1 ) otheriwse.

Reply:
I updated the constraints. Please take a look.